Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java

```diff
@@ -4143,7 +4143,7 @@ public boolean choose(Outcome outcome, Cards cards, TargetCard target, Ability s
                             continue;
                         }
                         if (hasObjectTargetNameOrAlias(card, targetName)) {
-                            if (target.isNotTarget() || target.canTarget(card.getId(), source, game)) {
+                            if (target.canTarget(getId(), card.getId(), source, game)) {
```

Comment:
There are three places with such code, not one.

I think it can be used to setup invalid targets for some reasons? If true then that's wrong usage and must be fixed in broken tests.

![shot_240531_140410](https://github.com/magefree/mage/assets/8344157/9c29fecd-4ed2-4d0f-96ed-882515e4b0e5)

Reply:
It's looks like a complex "problem" -- `isNotTarget || canTarget` used all around xmage:

![shot_240531_141532](https://github.com/magefree/mage/assets/8344157/005a3a99-9f5b-4b64-b2c5-e7aac44c39c7)

Reply:
I don't think there is an issue for real gameplay (we have choices using target code with the notTarget set to true, and that's important to be allowed to sacrifice a shroud creature for instance).

The thing with the test is that we were wrongly accepting some invalid choices. I'm running all the tests with the others 2 instances in TestPlayer without the call to isNotTarget(). I do believe this was wrong usage there and skipped validation.

Reply:
Game engine code is fine -- it's used for checking replacement effects (only real target :upside_down_face: can be restricted by effects):
![shot_240531_142206](https://github.com/magefree/mage/assets/8344157/1c7f9c88-7b54-4131-943a-016ef0c837bf)

I think `TestPlayer` must search targets same way. Example:

![shot_240531_142743](https://github.com/magefree/mage/assets/8344157/03d40dea-c8d2-4b07-ace0-f1d94a2038ad)

Reply:
That's protection effects can be broken in current test framework (it allow to add restricted target):

![shot_240531_144057](https://github.com/magefree/mage/assets/8344157/3657bb55-657a-41c3-aed2-98b202822a2b)